double_linked_list/double_linked_list.c

```diff
@@ -0,0 +1,153 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "double_linked_list.h"
+
+list* list_create(){
+    list* l = (list*)malloc(sizeof(list));
+    if (l == NULL)
+        return NULL;
+    l->quantity = 0;
+    l->head = NULL;
+    l->tail = NULL;
+    return l;
+}
+
+void list_free(list* l) {
+    if (l->head == NULL)
+        free(l);
+    else {
+        node *first = l->head;
+        if (l->quantity == 1)
+            free(first);
+        else if (l->quantity == 0)
+            return;
+        else {
+            while (first) {
+                node *second = first;
+                first = first->next;
+                free(second);
+            }
+        }
+        free(l);
+    }
+}
+
+void list_print(list* l){
+    if (!l)
+        printf("Empty list!\n");
+    else{
+        node* tmp = l->head;
+        printf("[ ");
+        while (tmp){
+            printf("%d ", tmp->data);
+            tmp = tmp->next;
+        }
+        printf("]\n");
+    }
+}
+
+bool list_prepend(list* l, int new_elem) {
+    node* new_node = (node*)malloc(sizeof(node));
+    if (!new_node)
+        return false;
+    new_node->data = new_elem;
+    new_node->next = l->head;
+    new_node->prev = NULL;
+    if (l->head)
+        l->head->prev = new_node;
+    l->head = new_node;
+    if (!l->head->next)
+        l->tail = l->head;
+    l->quantity += 1;
+    return true;
```

Comment:
Consider list, where `list->head` and `list->tail` are equal, and points to `node0`:

```
node0 [HEAD, TAIL]:
+------+------+------+
| PREV | DATA | NEXT |
+------+------+------+
| NULL |   0  | NULL |
+------+------+------+
```

Going through the code, we have:
1. New node allocation, initialization.
2. Interesting code:
```
    if (l->head)
        l->head->prev = new_node;
    l->head = new_node;
    if (!l->head->next)
        l->tail = l->head;
```
After firing it on above list example, both the `list->tail` and `list->head` points to the first node (which is `new_node`).